executor/executor.go

```diff
@@ -134,40 +133,76 @@ func Run(path string, env []string, emitter screwdriver.Emitter, build screwdriv
 
 	f.Write([]byte(shargs))
 
+	teardownFlag := false
+	var firstError error
+	var code int
+	var cmdErr error
+
 	cmds := build.Commands
 
 	for _, cmd := range cmds {
 		if err := api.UpdateStepStart(buildID, cmd.Name); err != nil {
 			return fmt.Errorf("Updating step start %q: %v", cmd.Name, err)
 		}
 
-		// Create step script file
-		stepFilePath := "/tmp/step.sh"
-		if err := createShFile(stepFilePath, cmd); err != nil {
-			return fmt.Errorf("Writing to step script file: %v", err)
-		}
+		isTeardown, _ := regexp.MatchString("sd-teardown-*", cmd.Name)
+
+		if !isTeardown {
+			// Generate guid for the step
+			guid := uuid.NewV4().String()
+
+			// Set current running step in emitter
+			emitter.StartCmd(cmd)
+			fmt.Fprintf(emitter, "$ %s\n", cmd.Cmd)
+
+			fReader := bufio.NewReader(f)
+
+			// Create step script file
+			stepFilePath := "/tmp/step.sh"
+
+			if err := createShFile(stepFilePath, cmd); err != nil {
+				return fmt.Errorf("Writing to step script file: %v", err)
+			}
 
-		// Generate guid for the step
-		guid := uuid.NewV4().String()
+			code, cmdErr = doRunCommand(guid, stepFilePath, emitter, f, fReader)
+		} else {
+			if !teardownFlag {
+				// Kill shell
+				exec.Command("exit")	// Need to check if the shell already exits
 
-		// Set current running step in emitter
-		emitter.StartCmd(cmd)
-		fmt.Fprintf(emitter, "$ %s\n", cmd.Cmd)
+				// Export environment variables
+				var exportCmd []string
 
-		fReader := bufio.NewReader(f)
+				for _, e := range env {
+					exportCmd = append(exportCmd, "export " + e)
+				}
+				exportString := strings.Join(exportCmd, " && ")
+
+				exec.Command(exportString)
+
+				// set teardownFlag so this block won't run again
+				teardownFlag = true
+			}
+
+			exec.Command(cmd.Cmd)
```

Comment:
This whole part should be something similar to this:
https://github.com/screwdriver-cd/launcher/blob/7be81506e9b1e34a2cab336b85fd1509bd231363/executor/executor.go#L32
We can probably just wrap in inside a separate function.

Reply:
Cool! 👍